Update library testing docs page to reduce confusion

noahfalk opened the pull request.

noahfalk committed:
Add warning on unmaintained testing doc page

stephentoub commented on docs/workflow/testing/libraries/testing.md:
```diff
@@ -1,5 +1,8 @@
 # Testing Libraries
 
+**WARNING** Instructions on this page appear to be unmaintained and do not work as written. See [here](https://github.com/dotnet/runtime/issues/54323) for details.
```
If this is true we should just delete it :)

noahfalk replied:
I'm fine with either of those options but I didn't want to presume. This felt like the minimal change that would warn others not to rely on it while we decided the fate of this page.

noahfalk committed:
Update testing.md

Some example text that seems more clear to me, but only offered as a suggestion. 
Feel free to adjust it or if you want to use it as-is please double check what I wrote is accurate : )

I think the useful elements are:
1. Being explicit about what workflow steps need to happen in total
2. Being explicit about which commands are covering the entire workflow and which ones are only covering a part of it
3. Show the simple "do-it-all" options first before showing more complex partial options. Glancing at the first example and blindly copying it should land in the pit of success.

ViktorHofer commented on docs/workflow/testing/libraries/testing.md:
```diff
@@ -1,47 +1,62 @@
+Before tests can run the following steps must happen:
+1. Build a runtime (either Mono or coreclr)
+2. Build the libraries
+3. Testhost deployment: Binaries must be arranged into a directory layout that tools/tests expect
```
Please remove this step as it should be transparent to the user. No action needs to be taken for this to happen.

noahfalk replied:
Leaving this step out entirely feels misleading. The [other docs page](https://github.com/dotnet/runtime/blob/8b116050be5735be8c94dcb8b0cd7554cab1e8dc/docs/workflow/building/libraries/README.md#iterating-on-systemprivatecorelib-changes) shows that explicit awareness and action is required in some some workflows:

> When changing System.Private.CoreLib after a full build, in order to test against those changes, you will need an updated System.Private.CoreLib in the testhost. In order to achieve that, you can build the libs.pretest subset which does testhost setup including copying over System.Private.CoreLib.

Lower on this page there is also explicit mention of running tests on individual assemblies which suggests that people reading this page are likely to be doing the single assembly build workflows where handling the deployment step manually is required.

I could probably restructure this and explain that it happens automatically when building all libs, but it does not happen automatically when rebuilding a single lib. Alternately we could divide up the content some other way?

ViktorHofer commented on docs/workflow/testing/libraries/testing.md:
````diff
@@ -1,47 +1,62 @@
+- Run all tests using clr:
 ```
-build.cmd/sh -subset libs.tests
+build.cmd/sh -subset clr+libs -test
````
```suggestion
build.cmd/sh -subset clr+libs+libs.tests -test
```

ViktorHofer replied:
Can you please double check that command actually works? I believe it's missing the `-restore` and `-build` arguments. You should be able to test that from a clean build.

noahfalk replied:
In the old docs it said "Unless you specifiy `-testnobuild`, test assemblies are implicitly built when invoking the `Test` action" which I took to mean that "libs.tests" was not required here. Was that inaccurate or I misunderstood?

ViktorHofer replied:
That is true but both `clr` and `libs` don't react to the `-test` action.

noahfalk replied:
Ah, that definitely mislead me but I think I understand now. I'll update accordingly.

ViktorHofer commented on docs/workflow/testing/libraries/testing.md:
```diff
@@ -1,47 +1,62 @@
+
+These examples automate all pre-requisite steps and the test run in a single command from a clean enlistment.
+
+- Run all tests using clr:
```
```suggestion
- Build all pre-requisites (clr + libs) and tests and run them:
```

ViktorHofer commented on docs/workflow/testing/libraries/testing.md:
````diff
@@ -1,47 +1,62 @@
 ```
 
-- The following builds and runs all tests using clr:
+- Run all tests using mono:
````
```suggestion
- Build all pre-requisites (mono + libs) and tests and run them using mono:
```

ViktorHofer commented on docs/workflow/testing/libraries/testing.md:
````diff
@@ -1,47 +1,62 @@
+- Run all tests using mono:
 ```
-build.cmd/sh -subset clr+libs.tests -test
+build.cmd/sh -subset mono+libs -test
````
```suggestion
build.cmd/sh -subset mono+libs+libs.tests -test
```

ViktorHofer commented on docs/workflow/testing/libraries/testing.md:
````diff
@@ -1,47 +1,62 @@
+- Run all tests - Builds clr in release, libs+tests in debug:
 ```
-build.cmd/sh -subset mono+libs.tests -test
+build.cmd/sh -subset clr+libs -test -rc Release
````
```suggestion
build.cmd/sh -subset clr+libs+libs.tests -test -rc Release
```

ViktorHofer commented on docs/workflow/testing/libraries/testing.md:
````diff
@@ -1,47 +1,62 @@
+- Run all tests - Build mono and libs for x86 architecture:
 ```
-build.cmd/sh -subset libs.tests -test -c Release
+build.cmd/sh -subset mono+libs -test -arch x86
````
```suggestion
build.cmd/sh -subset mono+libs+libs.tests -test -arch x86
```

noahfalk committed:
Apply suggestions from code review

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

noahfalk committed:
Update testing.md

noahfalk commented on docs/workflow/testing/libraries/testing.md:
````diff
@@ -1,54 +1,81 @@
-build.cmd/sh -subset mono+libs+libs.tests -test -arch x86
+build.cmd/sh -subset libs.pretest
 ```
+TODO: Is this a special case only for S.P.C or it is required for all libraries?
````
Unsure if what I wrote was accurate?

noahfalk commented on docs/workflow/testing/libraries/testing.md:
````diff
@@ -1,54 +1,81 @@
+- Build all pre-requisites (clr + libs) and tests and run them using CoreCLR:
 ```
-build.cmd/sh -subset libs.tests
+build.cmd/sh -subset clr+libs+libs.tests -test
````
I cloned a brand new repo and ran only this command. It is still in progress, but it appears to running successfully without additional arguments such as -build or -restore.

ViktorHofer replied:
Yes, just realized that -restore -build is implicit when passing in -test: https://github.com/dotnet/runtime/blob/f891033db5b8ebf651176a3dcc3bec74a217f85e/eng/build.ps1#L222-L224. It would not run if you also pass in `-testnobuild`.

noahfalk commented on docs/workflow/testing/libraries/testing.md:
````diff
@@ -1,54 +1,81 @@
 ```
 
-- The following builds mono and libs for x86 architecture and runs all tests:
+Building the `libs` subset automatically copies all product binaries into a testhost folder in the bin directory.
````
I restructured the content on the page to hopefully put this deployment step into better context. Let me know if this still appears to be surfacing unnecessary detail.

noahfalk committed:
Update testing.md

noahfalk committed:
Update testing.md

noahfalk committed:
Update testing.md

ViktorHofer commented on docs/workflow/testing/libraries/testing.md:
```diff
@@ -1,54 +1,80 @@
 
-- The following example shows how to pass extra msbuild properties to ignore tests ignored in CI:
+Building the `libs` subset automatically copies all product binaries into a testhost folder in the bin directory.
+This is where the tests will load the binaries from during the run. If you later [rebuild any of product binaries](https://github.com/dotnet/runtime/blob/main/docs/workflow/building/libraries/README.md#iterating-on-systemprivatecorelib-changes)
+using the individual library build projects you must also build the `libs.pretest` subset to copy the new binaries to testhost.
```
This isn't true. When building individual libraries, their output is copied into the testhost folder automatically. This infrastructure piece is called `binplacing`.

ViktorHofer replied:
That's why I expressed concerns about exposing the `libs.pretest` subset in this doc.

noahfalk replied:
Is the [existing documentation](https://github.com/dotnet/runtime/blob/8b116050be5735be8c94dcb8b0cd7554cab1e8dc/docs/workflow/building/libraries/README.md#iterating-on-systemprivatecorelib-changes) about incremental rebuild of S.P.C requiring libs.pretest inaccurate?

I can gladly update this but I am still attempting to distinguish between (a) it is a special case for S.P.C or (b) it isn't necessary in any circumstance.

ViktorHofer replied:
Right, CoreLib is special cased and doesn't binplace today. That should be fixed at some point but a concern was that its configuration need to be in sync with the native runtime bits.

cc @jkotas

noahfalk replied:
Hopefully the text now makes it clear that S.P.C is the special case and for everything else this is handled automatically.

noahfalk committed:
Update testing.md

noahfalk committed:
Update testing.md

noahfalk committed:
Update testing.md

jkotas commented on docs/workflow/testing/libraries/testing.md:
````diff
@@ -1,54 +1,76 @@
-**Examples**
-- The following shows how to build only the tests but not run them:
+- Build all pre-requisites (clr + libs) and tests and run them using CoreCLR:
 ```
````
This will build debug CoreCLR flavor that is super slow. `-rc Release` should be the first and recommended option here. It is what most libraries devs use.

noahfalk replied:
Done, examples with Release at the top and warning added on the debug example indicating it is very slow.

noahfalk committed:
Update testing.md

noahfalk committed:
Update testing.md

ViktorHofer commented:
@noahfalk thanks a lot for cleaning the docs up 👍

noahfalk commented:
@ViktorHofer np! Thanks for answering all the questions and reviewing it : )
